src/app/profile/my-spaces/my-spaces.component.ts

```diff
@@ -90,33 +145,97 @@ export class MySpacesComponent implements OnDestroy, OnInit {
       info: 'Start by creating a space.'
     } as EmptyStateConfig;
 
+    this.sharedSpacesEmptyStateConfig = {
+      title: 'No Spaces Found',
+      info: 'You are not listed as a collaborator of any spaces.',
+      helpLink: {
+        text: 'Find a Space to determine the owner of a space, then contact the space owner to request being added as a collaborator'
+      }
+    } as EmptyStateConfig;
+
     this.listConfig = {
       dblClick: false,
-      emptyStateConfig: this.emptyStateConfig,
+      emptyStateConfig: this.mySpacesEmptyStateConfig,
       headingRow: true,
       multiSelect: false,
       selectItems: false,
       showCheckbox: false,
       usePinItems: true
     } as ListConfig;
+  }
+
+  ngOnDestroy(): void {
+    this.subscriptions.forEach(sub => {
+      sub.unsubscribe();
+    });
+  }
+
+  initSpaces(): void {
+    if (this.context && this.context.user) {
+      this.subscriptions.push(
+        this.getMySpaces()
+          .pipe(
+            tap((mySpaces: Space[]): void => {
+              this._spaces = this.mySpaces = mySpaces;
+              this.updateSpaces();
+            }),
+            zip(this.getRoledSpaces()),
+            // find the difference between roledSpaces and mySpaces, if any - these IDs belong to shared spaces
+            map(([mySpaces, roledSpaces]: [Space[], SpaceInformation[]]): string[] => {
+              let roledSpaceIds: string[] = roledSpaces.map((space: SpaceInformation): string => space.id);
+              let mySpaceIds: string[] = mySpaces.map((space: Space): string => space.id);
+              return roledSpaceIds.filter((id: string) => mySpaceIds.indexOf(id) < 0);
+            }),
+            filter((ids: string[]): boolean => ids.length > 0),
+            switchMap((ids: string[]): Observable<Space[]> =>
+              forkJoin(ids.map((id: string): Observable<Space> => this.spaceService.getSpaceById(id)))
+            )
+          )
```

Comment:
Could all this zip/map/filter/switch logic be extracted out of the component and into an injected service? This is pretty heavyweight to do within a component, and would be more maintainable and testable in a service.

I would also prefer to avoid "taps with a side effect", so if this chain could be split up into multiple small chained Observables rather than one Subscription with a tapped side effect it would be nice.

Reply:
Sounds good, I'll move it into the `user-spaces.service` along side the other related functions.